### PR TITLE
Add basic user and session concepts, login form

### PR DESCRIPTION
**Pull request by Fnux**

### CODE REVIEW

File: app/src/main/java/ch/epfl/sweng/eventmanager/repository/UserRepository.java
Comment (zyuiop) [inline app/src/main/java/ch/epfl/sweng/eventmanager/repository/UserRepository.java]: What in this method isn't compatible with older APIs?

Comment (Fnux) [reply app/src/main/java/ch/epfl/sweng/eventmanager/repository/UserRepository.java]: It looks like streams and `Optional.ofNullable/1` require API 24 (current minimum is 15, see #42). I didn't dig in the documentation for details, this PR is still WIP.

**Comment by Fnux** [comment]
@zyuiop the current way to access `InMemorySession` singleton class is rather simple. We could move it to its own package, create a separate class for login/logout interactions, and use depency injection to load the session object in the various activities.

It add some complexity but I think it makes more sense given our current architecture.

**Comment by zyuiop** [comment]
Yes, I don't think it makes sense to have the login/logout mechanic directly in the InMemorySession.

**Comment by Fnux** [comment]
Apart from the support of older android API revisions and Travis failing on the testsfor some unknown reason, it's ready to be reviewed.

File: app/src/main/java/ch/epfl/sweng/eventmanager/ui/userManager/DisplayAccountActivity.java
Comment (zyuiop) [inline app/src/main/java/ch/epfl/sweng/eventmanager/ui/userManager/DisplayAccountActivity.java]: Session.getCurrentUser().get() might be null here (there are probably some cases where the user can glitch back to this activity after logging out)

File: app/src/main/java/ch/epfl/sweng/eventmanager/ui/userManager/LoginActivity.java
Comment (zyuiop) [inline app/src/main/java/ch/epfl/sweng/eventmanager/ui/userManager/LoginActivity.java]: Avoid hardcoded strings, put them in resources instead

File: app/src/main/java/ch/epfl/sweng/eventmanager/ui/userManager/LoginActivity.java
Comment (zyuiop) [inline app/src/main/java/ch/epfl/sweng/eventmanager/ui/userManager/LoginActivity.java]: Avoid hardcoded strings, put them in resources instead

File: app/src/main/java/ch/epfl/sweng/eventmanager/ui/userManager/LoginActivity.java
Comment (zyuiop) [inline app/src/main/java/ch/epfl/sweng/eventmanager/ui/userManager/LoginActivity.java]: Avoid hardcoded strings, put them in resources instead

File: app/src/main/java/ch/epfl/sweng/eventmanager/ui/eventSelector/EventPickingActivity.java
Comment (zyuiop) [inline app/src/main/java/ch/epfl/sweng/eventmanager/ui/eventSelector/EventPickingActivity.java]: Avoid hardcoded strings, put them in resources instead

File: app/src/main/java/ch/epfl/sweng/eventmanager/ui/eventSelector/EventPickingActivity.java
Comment (zyuiop) [inline app/src/main/java/ch/epfl/sweng/eventmanager/ui/eventSelector/EventPickingActivity.java]: Avoid hardcoded strings, put them in resources instead

File: app/src/main/java/ch/epfl/sweng/eventmanager/ui/userManager/LoginActivity.java
Comment (zyuiop) [inline app/src/main/java/ch/epfl/sweng/eventmanager/ui/userManager/LoginActivity.java:123]: Don't think it's a good idea, as the perfect email validation regex doesn't exist.

File: app/src/main/java/ch/epfl/sweng/eventmanager/ui/userManager/LoginActivity.java
Comment (zyuiop) [inline app/src/main/java/ch/epfl/sweng/eventmanager/ui/userManager/LoginActivity.java]: Take a `boolean` instead

File: app/src/main/java/ch/epfl/sweng/eventmanager/ui/userManager/LoginActivity.java
Comment (zyuiop) [inline app/src/main/java/ch/epfl/sweng/eventmanager/ui/userManager/LoginActivity.java]: I'd rather do 
```java
mLoginButton.setEnabled(!displayed);
mProgressBar.setVisibility(enabled ? View.VISIBLE : View.INVISIBLE);
```

It's still very readable and avoids what looks like "duplicating code" (`setEnabled` and `setVisibility` appear twice). That's just an opinion though.

File: app/src/main/java/ch/epfl/sweng/eventmanager/ui/userManager/LoginActivity.java
Comment (zyuiop) [inline app/src/main/java/ch/epfl/sweng/eventmanager/ui/userManager/LoginActivity.java]: Yep, we probably want to `@Inject` it in the activity (no need for a view model here I guess)

File: app/src/main/java/ch/epfl/sweng/eventmanager/userManagement/InMemorySession.java
Comment (zyuiop) [inline app/src/main/java/ch/epfl/sweng/eventmanager/userManagement/InMemorySession.java:11]: It works for now, in the future we might want to use @Inject directly

**Review by zyuiop — CHANGES_REQUESTED** [review]
Regarding the "string litterals" I mentioned, you already defined most of them in strings.xml but still use the litterals directly, you probably just forgot to replace them.

File: app/src/main/java/ch/epfl/sweng/eventmanager/ui/userManager/LoginActivity.java
Comment (Fnux) [reply app/src/main/java/ch/epfl/sweng/eventmanager/ui/userManager/LoginActivity.java]: Solved by d6ea85087bca781bb62f68fa1b14b1b87298664e.

**Comment by zyuiop** [comment]
Also, I tested the login activity, and it could be cool if using the "enter" icon in the integrated keyboard triggered the login the same way the button does (don't know how to handle that sadly)

File: app/src/main/java/ch/epfl/sweng/eventmanager/ui/userManager/LoginActivity.java
Comment (Fnux) [reply app/src/main/java/ch/epfl/sweng/eventmanager/ui/userManager/LoginActivity.java]: Solved by 2e9cd1a6a68ade17019b121fe3732e023029deb2.

File: app/src/main/java/ch/epfl/sweng/eventmanager/ui/userManager/LoginActivity.java
Comment (zyuiop) [reply app/src/main/java/ch/epfl/sweng/eventmanager/ui/userManager/LoginActivity.java:123]: (actually not a proper review, I close)

File: app/src/main/java/ch/epfl/sweng/eventmanager/userManagement/InMemorySession.java
Comment (zyuiop) [reply app/src/main/java/ch/epfl/sweng/eventmanager/userManagement/InMemorySession.java:11]: Not a proper review, I close

File: app/src/main/java/ch/epfl/sweng/eventmanager/ui/userManager/LoginActivity.java
Comment (Fnux) [reply app/src/main/java/ch/epfl/sweng/eventmanager/ui/userManager/LoginActivity.java:123]: The check was generated form an AndroidStudio template. I did not think of a perfect regex but more of the "general format" validation.

File: app/src/main/java/ch/epfl/sweng/eventmanager/userManagement/InMemorySession.java
Comment (Fnux) [reply app/src/main/java/ch/epfl/sweng/eventmanager/userManagement/InMemorySession.java:11]: I'd like to keep things as simple as possible - and we don't need dependency injection here for now.

File: app/src/main/java/ch/epfl/sweng/eventmanager/ui/userManager/DisplayAccountActivity.java
Comment (Fnux) [reply app/src/main/java/ch/epfl/sweng/eventmanager/ui/userManager/DisplayAccountActivity.java]: We have to implement a generic permission system instead of a specific validation. This will have to be on another sprint.

File: app/src/main/java/ch/epfl/sweng/eventmanager/ui/userManager/LoginActivity.java
Comment (zyuiop) [reply app/src/main/java/ch/epfl/sweng/eventmanager/ui/userManager/LoginActivity.java]: I'll find a solution for this

Comment (zyuiop) [reply app/src/main/java/ch/epfl/sweng/eventmanager/ui/userManager/LoginActivity.java]: I finally think it makes sense to have the repository directly in the Activity, as the task depends on the activity, and we might find some other uses for this repository directly in the activity in the future (actually, the task should probably be out of the activity and managed by the repository itself, but we will see once we will implement authentication against real async services)

**Review by zyuiop — DISMISSED** [review]

File: app/src/main/java/ch/epfl/sweng/eventmanager/ui/userManager/UserManagerModule.java
Comment (Fnux) [inline app/src/main/java/ch/epfl/sweng/eventmanager/ui/userManager/UserManagerModule.java]: What is this `import` doing here?

Comment (zyuiop) [reply app/src/main/java/ch/epfl/sweng/eventmanager/ui/userManager/UserManagerModule.java]: That's a fail

**Review by zyuiop — DISMISSED** [review]
Everything is ok for me!

**Comment by Fnux** [comment]
Note that this PR will switch back the minimal SDK version to 15 a tweak some of codeclimate's checks.

**Comment by Fnux** [comment]
I will merge this PR as soon as I get approval.

**Review by zyuiop — DISMISSED** [review]

**Review by zyuiop — DISMISSED** [review]

**Review by zyuiop — APPROVED** [review]

**Pull request merged**
